cibuildwheel/macos.py

```diff
@@ -49,37 +51,65 @@ def call(args, env=None, cwd=None, shell=False):
             print('+ ' + ' '.join(shlex_quote(a) for a in args))
 
         return subprocess.check_call(args, env=env, cwd=cwd, shell=shell)
-
-    # get latest pip once and for all
-    download(get_pip_url, get_pip_script)
-
-    for config in python_configurations:
-        # if this version of python isn't installed, get it from python.org and install
-        python_package_identifier = 'org.python.Python.PythonFramework-%s' % config.version
+  
+    def install_cpython(version, url):
+         # if this version of python isn't installed, get it from python.org and install
+        python_package_identifier = 'org.python.Python.PythonFramework-{}'.format(version)
         if python_package_identifier not in installed_system_packages:
             # download the pkg
-            download(config.url, '/tmp/Python.pkg')
+            download(url, '/tmp/Python.pkg')
             # install
             call(['sudo', 'installer', '-pkg', '/tmp/Python.pkg', '-target', '/'])
             # patch open ssl
-            if config.version == '3.5':
+            if version == '3.5':
                 open_ssl_patch_url = 'https://github.com/mayeut/patch-macos-python-openssl/releases/download/v1.0.2t/patch-macos-python-%s-openssl-v1.0.2t.tar.gz' % config.version
                 download(open_ssl_patch_url, '/tmp/python-patch.tar.gz')
-                call(['sudo', 'tar', '-C', '/Library/Frameworks/Python.framework/Versions/%s/' % config.version, '-xmf', '/tmp/python-patch.tar.gz'])
+                call(['sudo', 'tar', '-C', '/Library/Frameworks/Python.framework/Versions/{}/'.format(version), '-xmf', '/tmp/python-patch.tar.gz'])
+        return '/Library/Frameworks/Python.framework/Versions/{}/bin'.format(version)
+    
+    def install_pypy(version, url):
+        pypy_tar_bz2 = url.rsplit('/', 1)[-1]
+        assert pypy_tar_bz2.endswith(".tar.bz2")
+        pypy_base_filename = os.path.splitext(os.path.splitext(pypy_tar_bz2)[0])[0]
+        installation_path = os.path.join('/tmp', pypy_base_filename)
+        if not os.path.exists(installation_path):
+            download(url, os.path.join("/tmp", pypy_tar_bz2))
+            call(['tar', '-C', '/tmp', '-xf', os.path.join("/tmp", pypy_tar_bz2)])
+        return os.path.join(installation_path, 'bin')
+
+
+    # get latest pip once and for all
+    download(get_pip_url, get_pip_script)
+
+    for config in python_configurations:
+        if config.identifier.startswith('cp'):
```

Comment:
Maybe it also should be moved to function `upgrade_symlinks`?

Reply:
Wait, I don't understand what do you mean? This is just the installation, and afterwards determining the `python` and `pip` paths. I'm not upgrading symlinks, am I?

Reply:
from here to line 112 there are post installation step connected with calculating names and upgrade symlinks to python executables. Maybe all this things should be hide in install functions?

Reply:
But these are calculating variables we need later? But I'll have a look; maybe there's something we can do.

Reply:
This variables can be returned from function as tuple. In general this is big block of code which do one semantic thing. For me it is better to extract it because it simplify main function.

Reply:
I'm looking at it